server/routerlicious/packages/routerlicious-base/src/utils/sessionHelper.ts

```diff
@@ -267,9 +278,22 @@ export async function getSession(
 	documentRepository: IDocumentRepository,
 	sessionStickinessDurationMs: number = defaultSessionStickinessDurationMs,
 	messageBrokerId?: string,
+	clusterDrainingChecker?: IClusterDrainingChecker,
 ): Promise<ISession> {
 	const lumberjackProperties = getLumberBaseProperties(documentId, tenantId);
 
+	// Reject get session request if cluster is in draining process.
+	if (clusterDrainingChecker) {
+		const isClusterDraining = await clusterDrainingChecker.isClusterDraining();
+		if (isClusterDraining) {
+			Lumberjack.info("Cluster is in draining process. Reject get session request.");
+			throw new NetworkError(
+				503,
+				"Server is unavailable. Please retry session discovery later.",
+			);
+		}
```

Comment:
Why do we want to fail the request? even if the cluster is in draining, it can still provide the session info, right?

Reply:
I find that sometimes clients joined on old session still try to connect to cluster under draining. It could mess up the session status. For example, clients keep call get session using cluster's URL, not group endpoint. And once Deli updated the session flags, next call targeting the old cluster will mark the session active on old cluster again (if the old cluster handles the request). I am not sure why it happens for now, so I add this guard to prevent drained cluster from handling get session API.

Reply:
I'm with Yang here. Could we move this guard to below line 303 (now line 330) after `existingSession` check is confirmed? As long as we don't _create_ a new session in the drained cluster, and the `existingSession` is not in this cluster, it would be ideal to redirect the session rather than fail it, right?

Reply:
It seems that it would be more ideal to fail with the 503 _after_ we at least check `if (existingSession && (existingSession.isSessionAlive || existingSession.isSessionActive)`. We could also add an extra check there to make sure `exisitingSession` is not in draining cluster. But overall, I think it would be preferable to redirect rather than fail if possible, right?

Reply:
Moved it after the check.